pkg/controller/traceflow/controller.go

```diff
@@ -0,0 +1,306 @@
+	c.enqueueTraceflow(tf)
+}
+
+func (c *Controller) deleteTraceflow(old interface{}) {
+	tf := old.(*opsv1alpha1.Traceflow)
+	klog.Infof("Processing Traceflow %s DELETE event", tf.Name)
+	c.deallocateTag(tf)
+}
+
+// worker is a long-running function that will continually call the processTraceflowItem function
+// in order to read and process a message on the workqueue.
+func (c *Controller) worker() {
+	for c.processTraceflowItem() {
+	}
+}
+
+// processTraceflowItem processes an item in the "traceflow" work queue, by calling syncTraceflow
+// after casting the item to a string (Traceflow name). If syncTraceflow returns an error, this
+// function logs error. If syncTraceflow returns retry flag is false, the Traceflow will be added
+// to queue with rate limit. If syncTraceflow returns retry flag is false, the Traceflow is removed
+// from the queue until we get notified of a new change. This function returns false if and only if
+// the work queue was shutdown (no more items will be processed).
+func (c *Controller) processTraceflowItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// We call Done here so the workqueue knows we have finished processing this item. We also
+	// must remember to call Forget if we do not want this work item being re-queued. For
+	// example, we do not call Forget if a transient error occurs, instead the item is put back
+	// on the workqueue and attempted again after a back-off period.
+	defer c.queue.Done(obj)
+
+	// We expect strings (Traceflow name) to come off the workqueue.
+	if key, ok := obj.(string); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen: enqueueTraceflow only enqueues strings.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected string in work queue but got %#v", obj)
+		return true
+	} else {
+		retry, err := c.syncTraceflow(key)
+		if err != nil {
+			klog.Errorf("Error syncing Traceflow %s, Aborting. Error: %v", key, err)
+		}
+		// Add key to queue if retry flag is true, forget key if retry flag is false.
+		if retry {
+			c.queue.AddRateLimited(key)
+		} else {
+			c.queue.Forget(key)
+		}
+	}
+	time.Sleep(time.Duration(1) * time.Second)
+	return true
+}
+
+func (c *Controller) syncTraceflow(traceflowName string) (bool, error) {
+	startTime := time.Now()
+	defer func() {
+		klog.V(4).Infof("Finished syncing Traceflow for %s. (%v)", traceflowName, time.Since(startTime))
+	}()
+
+	retry := false
+	tf, err := c.traceflowLister.Get(traceflowName)
+	if err != nil {
+		retry = true
+		return retry, err
+	}
+	switch tf.Status.Phase {
+	case "", opsv1alpha1.Pending:
+		_, err = c.startTraceflow(tf)
+		retry = true
+	case opsv1alpha1.Running:
+		if time.Now().UTC().Sub(tf.CreationTimestamp.UTC()).Seconds() > timeout {
+			_, err = c.errorTraceflowCRD(tf)
+		} else {
+			c.updateTagCache(tf)
```

Comment:
why it needs to updateTagCache when the traceflow doesn't timeout

Reply:
No need to update cache now, we changed cache to
runningTraceflows      map[uint8]string
Removed updateTagCache